stackslib/src/chainstate/nakamoto/miner.rs

```diff
@@ -833,3 +835,252 @@ impl BlockBuilder for NakamotoBlockBuilder {
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks is acceptable.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateOk {
+    pub block: NakamotoBlock,
+    pub cost: ExecutionCost,
+    pub size: u64,
+}
+
+/// This enum is used for serializing the response to block
+/// proposal validation.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "Result")]
+pub enum BlockValidateResponse {
+    Ok(BlockValidateOk),
+    Reject(BlockValidateReject),
+}
+
+impl From<Result<BlockValidateOk, BlockValidateReject>> for BlockValidateResponse {
+    fn from(value: Result<BlockValidateOk, BlockValidateReject>) -> Self {
+        match value {
+            Ok(o) => BlockValidateResponse::Ok(o),
+            Err(e) => BlockValidateResponse::Reject(e),
+        }
+    }
+}
+
+impl From<Error> for BlockValidateReject {
+    fn from(value: Error) -> Self {
+        BlockValidateReject {
+            reason: format!("Chainstate Error: {value}"),
+            reason_code: ValidateRejectCode::ChainstateError,
+        }
+    }
+}
+
+impl From<DBError> for BlockValidateReject {
+    fn from(value: DBError) -> Self {
+        Error::from(value).into()
+    }
+}
+
+/// Represents a block proposed to the `v2/block_proposal` endpoint for validation
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct NakamotoBlockProposal {
+    /// Proposed block
+    pub block: NakamotoBlock,
+    // tenure ID -- this is the index block hash of the start block of the last tenure (i.e.
+    // the data we committed to in the block-commit).  If this is an epoch 2.x parent, then
+    // this is just the index block hash of the parent Stacks block.
+    pub tenure_start_block: StacksBlockId,
+    /// Identifies which chain block is for (Mainnet, Testnet, etc.)
+    pub chain_id: u32,
+}
+
+impl StacksMessageCodec for NakamotoBlockProposal {
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), CodecError> {
+        write_next(fd, &self.block)?;
+        write_next(fd, &self.tenure_start_block)?;
+        write_next(fd, &self.chain_id)
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<Self, CodecError> {
+        Ok(Self {
+            block: read_next(fd)?,
+            tenure_start_block: read_next(fd)?,
+            chain_id: read_next(fd)?,
+        })
+    }
+}
+
+impl NakamotoBlockProposal {
+    /// Test this block proposal against the current chain state and
+    /// either accept or reject the proposal
+    ///
+    /// This is done in 2 steps:
+    /// - Static validation of the block, which checks the following:
+    ///   - Block is well-formed
+    ///   - Transactions are well-formed
+    ///   - Miner signature is valid
+    /// - Validation of transactions by executing them agains current chainstate.
+    ///   This is resource intensive, and therefore done only if previous checks pass
+    pub fn validate(
```

Comment:
This is duplicating lots of logic in `NakamotoChainState::process_next_nakamoto_block()`. Can you instead just make it so that method can optionally (1) take a particular block to process as an argument, and (2) just return the receipts _without_ committing the new chainstate?  Then, you can just call that directly instead of duplicating the same checks.

Reply:
Sure

Reply:
Pushed my work to `feat/nakamoto-block-proposal-refactor-validate` if you want to check it out (doesn't compile)

Reply:
Actually, thinking about this a bit more, I think what you have here is fine for now.  We'll refactor `NakamotoChainState::process_next_nakamoto_block()` so this code can crib from it without needing a `&mut SortitionDB`.  `NakamotoChainState::process_next_nakamoto_block()` has lots of places where it can write data, so it doesn't make sense to use it here as-is since this function is not meant to ever write data to the chainstate.  Also, this function shouldn't open a transaction against any DBs if we can help it.